### PR TITLE
gateway: Add requestContext arg to Experimental_DidResolveQueryPlanCa…

### PR DESCRIPTION
**Pull request by alex-lange**
We are hoping to add authorization checks based on the services called in the queryPlan, and adding such a check in this experimental callback seems like a good idea. We need the requestContext because that is where we are storing our credential information.

### CODE REVIEW

File: packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts [hunk 1]
```diff
@@ -217,4 +219,25 @@ describe('lifecycle hooks', () => {
...
+      document: gql`
+        { product(upc: '1') { name } }
+      `,
+      request: {},
+      queryHash: 'hashed',
+      context: {},
+    });
+
+    expect(experimental_didResolveQueryPlan).toBeCalled();
+  });
```
Comment (alex-lange) [inline packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts:242]: I didn't find any tests for `Experimental_DidResolveQueryPlanCallback`

File: packages/apollo-gateway/src/index.ts [hunk 2]
```diff
@@ -105,10 +105,12 @@ export type Experimental_DidResolveQueryPlanCallback = ({
   queryPlan,
   serviceMap,
   operationContext,
+  requestContext,
 }: {
   readonly queryPlan: QueryPlan;
   readonly serviceMap: ServiceMap;
   readonly operationContext: OperationContext;
+  readonly requestContext?: GraphQLRequestContextExecutionDidStart<Record<string, any>>;
```
Comment (alex-lange) [inline packages/apollo-gateway/src/index.ts:113]: I think the `?:` is needed for backwards compatibility? Also I'm not sure if `Record<string, any>` is right.

Comment (trevor-scheer) [reply packages/apollo-gateway/src/index.ts:113]: As long as `this.experimental_didResolveQueryPlan` is called with this object in all cases (it is, in its one usage below), then the optional `?` is not needed. The implementor isn't required to accommodate the new existence of this object, so it's a backwards-compatible change.

```suggestion
  readonly requestContext: GraphQLRequestContextExecutionDidStart<Record<string, any>>;
```

Comment (trevor-scheer) [reply packages/apollo-gateway/src/index.ts:113]: `Record<string, any>` is a fine compromise for now. The `context` typing throughout Apollo Server has some issues that I'd like to tackle, but they are breaking changes and have sweeping effects across many internal and externally facing interfaces and classes. I am happy with this for now.

**Comment by trevor-scheer** [comment]
Released in `@apollo/gateway@0.16.4`. Thanks for the contribution 🎉!

**Pull request merged**